proposals/nnnn-delayed-enqueuing.md

```diff
@@ -0,0 +1,269 @@
+
+* Proposal: [SE-NNNN](NNNN-scheduling-executor.md)
+* Authors: [Alastair Houghton](https://github.com/al45tair)
+* Review Manager: TBD
+* Status: **Pitch**
+* Implementation: On main branch
+* Review: ([first
+  pitch](https://forums.swift.org/t/pitch-custom-main-and-global-executors/77247))
+  ([second pitch](https://forums.swift.org/t/pitch-2-custom-main-and-global-executors/78437))
+  ([third pitch](https://forums.swift.org/t/pitch-3-custom-main-and-global-executors/80638))
+
+## Introduction
+
+(This was previously pitched as part of the Custom Main and Global
+Executors proposal, but has been split out into a separate proposal
+here.)
+
+The Swift Concurrency runtime provides some C entry points that allow
+code to schedule jobs on the global executor after a delay or at a
+point in time.  These entry points are not ideal; in particular:
+
+  * `swift_task_enqueueGlobalWithDelay` does not support tolerances,
+    will only use the continuous clock, and will only schedule on the
+    global executor
+
+  * `swift_task_enqueueGlobalWithDeadline` expresses the timestamp as
+    two `long long` values, one for seconds and one for nanoseconds,
+    using a similar format for the tolerance (or "leeway").  It only
+    supports the built-in clocks, and will only schedule on the global
+    executor.
+
+The `Executor` protocol already provides a means for Swift code to
+enqueue a job, and it makes sense that it should also provide a way
+for Swift code to enqueue with a delay or deadline.
+
+## Motivation
+
+As part of the work on custom executors, we will need to be able to
+implement the aforementioned entry points such that calling them will
+hand off to the user's custom executor implementation.
+
+We would also like to be able to support arbitrary `Clock`
+implementations, rather than being stuck with either just the
+`ContinuousClock`, or a choice of `ContinuousClock` and
+`SuspendingClock`.
+
+## Proposed solution
+
+Since there are extant custom `Executor` implementations, we can't
+easily extend `Executor` itself --- if we did so, the extension would
+have to be optional, and then there would need to be a way to discover
+that we didn't have the new methods.  It seems better to add a new
+protocol, `SchedulingExecutor`, that we can use to determine whether
+or not a given executor supports the new API surface:
+
+```swift
+protocol SchedulingExecutor: Executor {
+  ...
+  /// Enqueue a job to run after a specified delay.
+  ///
+  /// You need only implement one of the two enqueue functions here;
+  /// the default implementation for the other will then call the one
+  /// you have implemented.
+  ///
+  /// Parameters:
+  ///
+  /// - job:       The job to schedule.
+  /// - after:     A `Duration` specifying the time after which the job
+  ///              is to run.  The job will not be executed before this
+  ///              time has elapsed.
+  /// - tolerance: The maximum additional delay permissible before the
```

Comment:
What happens if the "impermissible" happens and the time has exceeded the tolerance given? Is the job then never run, or is it run as soon as possible? Is there some sort of error raised?

Reply:
The behaviour I'd expect from an executor implementation is that it will try to run jobs before the deadline (e.g. a simple strategy might be to run the jobs in deadline order), but never before *after*, that job scheduling will be best effort and that all jobs will execute. So maybe "permissible" is the wrong word.

I can imagine that there might be a use-case for a hard deadline, for instance if you have a real-time scheduler, but I think we would want to add more `enqueue` methods to support that kind of behaviour (for example, I think you'd want an estimate of the time the job could run for, and you'd want the `enqueue` method to return a `Bool` to tell you whether it had actually been able to schedule the job). That's outside of the scope of this work though, I think.